CRM/Contribute/BAO/Contribution.php

```diff
@@ -177,6 +177,11 @@ public static function add(&$params, $ids = array()) {
       $params['prevContribution'] = self::getValues(array('id' => $contributionID), CRM_Core_DAO::$_nullArray, CRM_Core_DAO::$_nullArray);
     }
 
+    // CRM-16189
+    $error = CRM_Financial_BAO_FinancialAccount::checkForValidFinancialType($params, $contributionID);
```

Comment:
this seems to be a new function so let's have it throw an exception. In general we prefer BAO functions (& lots of other things) to throw exceptions rather than return errors - which leads to ugly stuff like this. The form layer should catch & handle the exceptions